api/src/main/java/org/pmiops/workbench/db/dao/WorkspaceServiceImpl.java

```diff
@@ -0,0 +1,112 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import org.pmiops.workbench.db.model.Workspace;
+import org.pmiops.workbench.db.model.WorkspaceUserRole;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.exceptions.NotFoundException;
+
+/**
+ * Workspace manipulation and shared business logic which can't be represented by automatic query
+ * generation in WorkspaceDao, or convenience aliases.
+ *
+ * This needs to implement an interface to support Transactional
+ *
+ * TODO(RW-215) Add versioning to detect/prevent concurrent edits.
+ */
+@Service
+public class WorkspaceServiceImpl implements WorkspaceService {
+  private static final Logger log = Logger.getLogger(WorkspaceService.class.getName());
+
+  @Autowired private WorkspaceDao workspaceDao;
+
+  /**
+   * Clients wishing to use the auto-generated methods from the DAO interface may directly access
+   * it here.
+   */
+  @Override
+  public WorkspaceDao getDao() {
+    return workspaceDao;
+  }
+  @Override
+  public void setDao(WorkspaceDao workspaceDao) {
+    this.workspaceDao = workspaceDao;
+  }
+  @Override
+  public Workspace get(String ns, String id) {
+    return workspaceDao.findByWorkspaceNamespaceAndFirecloudName(ns, id);
+  }
+  @Override
+  public Workspace getRequired(String ns, String id) {
+    Workspace workspace = get(ns, id);
+    if (workspace == null) {
+      throw new NotFoundException(String.format("Workspace %s/%s not found.", ns, id));
+    }
+    return workspace;
+  }
+  @Override
+  public List<Workspace> findForReview() {
+    return workspaceDao.findByApprovedIsNullAndReviewRequestedTrueOrderByTimeRequested();
+  }
+
+  // FIXME @Version instead? Bean instantiation v. @Transactional?
+  //@Transactional
+  @Override
+  public void setResearchPurposeApproved(String ns, String id, boolean approved) {
+    Workspace workspace = getRequired(ns, id);
+    if (workspace.getReviewRequested() == null || !workspace.getReviewRequested()) {
+      throw new BadRequestException(String.format(
+          "No review requested for workspace %s/%s.", ns, id));
+    }
+    if (workspace.getApproved() != null) {
+      throw new BadRequestException(String.format(
+          "Workspace %s/%s already %s.",
+          ns, id, workspace.getApproved() ? "approved" : "rejected"));
+    }
+    workspace.setApproved(approved);
+    workspaceDao.save(workspace);
+  }
+
+  @Override
+  @Transactional
+  public void updateUserRoles(String ns, String id, Set<WorkspaceUserRole> userRoleSet) {
+    org.pmiops.workbench.db.model.Workspace dbWorkspace = getRequired(
+        ns, id);
+    Map<Integer, WorkspaceUserRole> userRoleMap = new HashMap<Integer, WorkspaceUserRole>();
+    for (WorkspaceUserRole userRole : userRoleSet) {
+      userRole.setWorkspace(dbWorkspace);
+      int key = Objects.hash(userRole.getUser().getUserId(), userRole.getWorkspace().hashCode());
```

Comment:
This could result in hash collisions. If you had to include both user and workspace ID, you should use a Pair or POJO of some kind.

But you don't actually need to include workspace ID in the key; it's going to be the same for all of these. Just make this a Map of user ID -> WorkspaceUserRole.